### PR TITLE
feat: add newAllocation request

### PR DESCRIPTION
**Pull request by kushthedude**
Signed-off-by: kushthedude <kushthedude@gmail.com>

Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/blobber/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master BEFORE merging this PR

Associated PRs (Link as appropriate):
...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | alloc_blobber  |
| zboxcli  | new_alloc  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | alloc_blobber  |
| zboxcli  | new_alloc  |
| zwalletcli  | staging  |
| 0chain  | alloc_blob  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | new_alloc  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | alloc_blobber  |
| zboxcli  | new_alloc  |
| zwalletcli  | staging  |
| 0chain  | alloc_blob  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

File: zboxcore/sdk/sdk.go [hunk 1]
```diff
@@ -886,6 +892,43 @@ func CreateAllocationForOwner(owner, ownerpublickey string,
...
+		"diversify_blobbers":            false,
+	}
+
+	allocationData, _ := json.Marshal(allocationRequest)
+
+	params := make(map[string]string)
+	params["allocation_data"] = string(allocationData)
+	params["max_challenge_time"] = mcct.String()
+
+	allocBlobber, err := zboxutil.MakeSCRestAPICall(STORAGE_SCADDRESS, "/get_alloc_blobbers", params, nil)
```
Comment (Sriep) [inline zboxcore/sdk/sdk.go:919]: The call to `get_alloc_blobbers` should pass exactly the parameters that are required.

-  number of blobbers required
- size of allocation
- expiration
- min read price
- max read price
- min write price
- max write price
- max challenge completion time

The number of blobbers required is to be calculated as
```
number of blobbers = data_shards + parity shards - number of preferred blobbers
```
If the number of preferred blobbers is >= data shards + parity shards, then we can dispense with the get_alloc_blobbers" call.

File: zboxcore/sdk/sdk.go [hunk 2]
```diff
@@ -886,6 +892,43 @@ func CreateAllocationForOwner(owner, ownerpublickey string,
...
+		"data_shards":                   datashards,
+		"parity_shards":                 parityshards,
+		"size":                          size,
+		"owner_id":                      owner,
+		"owner_public_key":              ownerpublickey,
+		"expiration_date":               expiry,
+		"read_price_range":              readPrice,
+		"write_price_range":             writePrice,
+		"max_challenge_completion_time": mcct,
+		"diversify_blobbers":            false,
```
Comment (Sriep) [inline zboxcore/sdk/sdk.go:911]: diversify_blobbers field should be removed. It is not needed any more..

File: zboxcore/sdk/sdk.go [hunk 3]
```diff
@@ -886,6 +892,43 @@ func CreateAllocationForOwner(owner, ownerpublickey string,
...
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to make SC-REST Call")
+	}
+
+	var allocBlobberIDs []string
+
+	err = json.Unmarshal(allocBlobber, &allocBlobberIDs)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal blobber IDs")
+	}
```
Comment (Sriep) [inline zboxcore/sdk/sdk.go:928]: There should b a call to `new_allocation_request` somewhere around here. We should add to `allocationRequest` parameters, a `blobbers` field which will be a concatenation of the preferred blobbers sent by the caller with the blobber ids returned by `get_alloc_blobbers`.

File: zboxcore/sdk/sdk.go [hunk 4]
```diff
@@ -871,7 +877,7 @@ func CreateAllocationForOwner(owner, ownerpublickey string,
 		"owner_id":                      owner,
 		"owner_public_key":              ownerpublickey,
 		"expiration_date":               expiry,
-		"preferred_blobbers":            preferredBlobbers,
+		"preferred_blobbers":            allocationBlobbers,
```
Comment (Sriep) [inline zboxcore/sdk/sdk.go]: Keep this as `preferredBlobbers`, `allocationBlobbers` can be the `preferredBlobbers` concatenated with the ids returned from `get_alloc_blobbers`

**Review by Sriep — CHANGES_REQUESTED** [review]

File: zboxcore/sdk/sdk.go (hunk 1 above)
Comment (kushthedude) [reply zboxcore/sdk/sdk.go:919]: @Sriep Yes, the number of parameters were same as what are required for the old allocation request, but the flow was not working, hence right now the parameters are just minimal to actually see if gosdk & zbox are able to get the blobbers & populate them as prefferedBlobbers.

File: zboxcore/sdk/sdk.go (hunk 4 above)
Comment (kushthedude) [reply zboxcore/sdk/sdk.go]: This is going to be removed once 684 gets in, prefferedBlobbers would be made Blobbers & the list of blobbers passed from zbox will be the prefferedBlobbers & then the blobbers fetched from eventsDB endpoint would be the desired blobbers.
Preffered + desired = dataParity + shardParity, this is the number of blobbers which we are gonna send to the new allocation request.

**Comment by dabasov** [comment]
duplicate #449

**Pull request closed without merging**